### PR TITLE
DOCPLAT-1447 Move cache, a few modifications

### PR DESCRIPTION
**Pull request by aaronstgeorge-wf**
### Description

A client side caching implementation is needed in a few places. This PR moves an existing cache from an internal Workiva repo with a few modifications.

### Changes

Moved cache from an inheritance to a strategy based pattern. Consumers may now use the same `Cache` API opaquely from the strategy used. Entirely generic strategies can be swapped in and out without needing to change any consumption code.

Removed the `put` method. This seemed like a nice to have, as a `remove` followed by a `get` could always accomplish the same thing. The semantics of what each future should resolve to on if a `getAsync('id', AsyncValueFactory)` `put('id', vlaue)` `getAsync('id' AsyncValueFactory)` came in synchronously with an a were hard to determine, whereas a `get` `remove` `get` is clear.

### Semantic Versioning

- [ ] **Patch**
  - [ ] This change does not affect the public API
  - [ ] This change fixes existing incorrect behavior without any additions
- [x] **Minor**
  - [x] This change adds something to the public API
  - [ ] This change deprecates a part of the public API
* [ ] Major
  - [ ] This change modifies part of the public API in a backwards-incompatible manner
  - [ ] This change removes part of the public API

### Testing/QA

- [ ] Open sourcing this is fine
- [ ] CI passes
- [ ] Try this in the internal Workiva app that it came from to ensure that it still functions properly 

### Code Review

@bencampbell-wf 
@trentgrover-wf

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Raven
<details>
<summary>Number of Findings: 0</summary>

</details>

File: lib/src/common/cache/cache.dart [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+  ///     var c = cache.getAsync('id', _superLongAsyncCall);
+  ///
+  ///     // The futures in this list will all resolve to the same instance,
+  ///     // which was returned by `_superLongAsyncCall` the first time `get` was
+  ///     // called.
+  ///     var values = Future.wait([a, b, c]);
+  ///
+  /// If the [Cache] [isOrWillBeDisposed] then a [StateError] is thrown.
+  @mustCallSuper
+  Future<TValue> getAsync(TIdentifier id, Func<Future<TValue>> valueFactory) {
```
Comment (aaronstgeorge-wf) [inline lib/src/common/cache/cache.dart]: `Get` and `GetAsync` could be combined by using `FutureOr`. I'm not sure how people feel about that

File: lib/src/common/cache/cache.dart [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+  ///         remove(TIdentifier id)) async {
+  ///       var timer = getManagedTimer(new Duration(seconds:30), () {
+  ///         remove(id);
+  ///       });
+  ///     }
+  ///
+  /// Values should not be removed unnecessarily. In the example below it is
+  /// expected that both `a`, `b` and `c` would resolve to the same value while
+  /// only making `_superLongAsyncCall` once. Here [Cache.release] and
+  /// [Cache.get] are called synchronously before `_superLongAsyncCall` has
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: `Cache.get` --> `Cache.getAsync`

File: lib/src/common/cache/cache.dart [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+  /// completed. The strategy has enough information to know that even though
+  /// release was called there is no need for the value to be evicted and
+  /// recomputed, (because of the get immediately after the release). Release is
+  /// a suggestion, every release does not need to be paired with a removal. If
+  /// consumers wanted `_superLongAsyncCall` in the example below to be ran
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: `to be run`

File: lib/src/common/cache/cache.dart [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+  /// the cache.
+  Stream<CacheContext<TIdentifier, TValue>> get didRemove =>
+      _didRemoveController.stream;
+
+  /// The stream of [CacheContext]s that dispatches when an item is updated in
+  /// the cache.
+  Stream<CacheContext<TIdentifier, TValue>> get didUpdate =>
+      _didUpdateController.stream;
+
+  /// Returns a value from the cache for a given [TIdentifier]. If the
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: 🌵 The first sentence of a doc comment should be a standalone paragraph.

File: lib/src/common/cache/cache.dart [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    }
+
+    // Install Future value
+    final Completer<TValue> completer = new Completer<TValue>();
+    _cache[id] = completer.future;
+    try {
+      final TValue value = valueFactory.call();
+
+      _cachingStrategy.onDidGet(id, value).then((Null _) {
+        _didUpdateController.add(new CacheContext(id, value));
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: In this case are we really adding the value to the cache? i.e. should this event be didAdd instead of didUpdate?

File: lib/src/common/cache/cache.dart [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+    } catch (error, stackTrace) {
+      completer.completeError(error, stackTrace);
+    }
+
+    return _cache[id];
+  }
+
+  /// Returns a value from the cache for a given [TIdentifier]. If the
+  /// [TIdentifier] is not present in the cache the value returned by the given
+  /// [valueFactory] is added to the cache and returned.
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: 🌵 The first sentence should be a standalone paragraph.

File: lib/src/common/cache/cache.dart (hunk 1 above)
Comment (todbachman-wf) [reply lib/src/common/cache/cache.dart]: Eh 🤷‍♂️

File: lib/src/common/cache/cache.dart [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+  ///
+  /// All calls to [getAsync] await a call to the current caching strategy's
+  /// [CachingStrategy.onDidGet] method before returning the cached value. If the
+  /// [Cache] does not contain an instance for the given [TIdentifier], the given
+  /// [valueFactory] is called and an [CacheContext] event is emitted on the
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: `is called and a`

File: lib/src/common/cache/cache.dart [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+  /// [didUpdate] stream. A call to [get] that returns a cached value does not
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: `[get]` --> `[getAsync]`

File: lib/src/common/cache/cache.dart [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+  /// emit this event as the [Cache] has not updated.
+  ///
+  /// Calls to [getAsync] are evaluated synchronously and will return the future
+  /// value established by the first call to [get]. This allows calls to
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: `[get]` --> `[getAsync]`

File: lib/src/common/cache/cache.dart [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+      return value;
+    });
+
+    return _cache[id];
+  }
+
+  Iterable<TIdentifier> get keys => _cache.keys;
+
+  /// Releases a reference for a given [TIdentifier] and removes the
+  /// [TIdentifier] from the [Cache] when the last reference is released.
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: This is true in the reference count cache, but not generically.

File: lib/src/common/cache/cache.dart [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+  @mustCallSuper
+  Future<Null> remove(TIdentifier id) {
+    _throwWhenDisposed('remove');
+    if (_cache.containsKey(id)) {
+      final removedValue = _cache.remove(id);
+      _cachingStrategy.onWillRemove(id);
+      return removedValue.then((TValue value) async {
+        await _cachingStrategy.onDidRemove(id, value);
+        _didRemoveController.add(new CacheContext(id, value));
+        _didUpdateController.add(new CacheContext(id, null));
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: Ah, I see we're using didUpdate here too. So maybe we don't want to change it to didAdd.

File: smithy.yaml [hunk 12]
```diff
@@ -5,7 +5,7 @@ runner_image: drydock-prod.workiva.net/workiva/smithy-runner-generator:203768 #
 
 script:
   - pub get --packages-dir
-  - xvfb-run -s '-screen 0 1024x768x24' pub run dart_dev test --pub-serve --web-compiler=dartdevc -p chrome -p vm
+  - xvfb-run -s '-screen 0 1024x768x24' pub run dart_dev test --pub-serve --web-compiler=dart2js -p chrome -p vm
```
Comment (todbachman-wf) [inline smithy.yaml]: Why the change?

File: test/unit/browser/cache/caching_strategy_common_test.dart [hunk 13]
```diff
@@ -0,0 +1,63 @@
+import 'dart:async';
+import 'package:test/test.dart';
```
Comment (todbachman-wf) [inline test/unit/browser/cache/caching_strategy_common_test.dart]: Might wanna organize imports. 😉

File: smithy.yaml (hunk 12 above)
Comment (aaronstgeorge-wf) [reply smithy.yaml]: I was getting some type errors from the dart dev compiler that didn't show up in analysis. As it turns out strong mode compliance != ddc compliance in a few annoying circumstances. Using `typed(any)` rather than `any` with mokito fixes the errors I was seeing however. https://github.com/dart-lang/mockito#strong-mode-compliance . So I'll be switching it back in an upcoming commit.

File: lib/src/common/cache/cache.dart (hunk 2 above)
Comment (todbachman-wf) [reply lib/src/common/cache/cache.dart]: Shouldn't that be changed to `getAsync`?

File: lib/src/common/cache/cache.dart [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+  ///       });
+  ///     }
+  ///
+  /// Values should not be removed unnecessarily. In the example below it is
+  /// expected that both `a`, `b` and `c` would resolve to the same value while
+  /// only making `_superLongAsyncCall` once. Here [Cache.release] and
+  /// [Cache.get] are called synchronously before `_superLongAsyncCall` has
+  /// completed. The strategy has enough information to know that even though
+  /// release was called there is no need for the value to be evicted and
+  /// recomputed, (because of the get immediately after the release). Release is
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: `recomputed,` --> `recomputed`

File: lib/src/common/cache/cache.dart [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+
+  Cache(this._cachingStrategy) {
+    [_didReleaseController, _didRemoveController, _didUpdateController]
+        .forEach(manageStreamController);
+  }
+
+  /// Does the [Cache] contain the given [TIdentifier]?
+  ///
+  /// If the [Cache] [isOrWillBeDisposed] then a [StateError] is thrown.
+  bool containsKey(Object id) {
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: Shouldn't the type of id be `TIdentifier`?

File: lib/src/common/cache/cache.dart [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+      });
+    }
+
+    // Install Future value
+    final Completer<TValue> completer = new Completer<TValue>();
+    _cache[id] = completer.future;
+    try {
+      final TValue value = valueFactory.call();
+
+      _cachingStrategy.onDidGet(id, value).then((Null _) {
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: Should we be awaiting this onDidGet? We say we do in the doc comment.

File: lib/src/common/cache/cache.dart (hunk 9 above)
Comment (todbachman-wf) [reply lib/src/common/cache/cache.dart]: Shouldn't we change that get to getAsync?

File: lib/src/common/cache/cache.dart [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+  /// Returns a value from the cache for a given [TIdentifier].
+  ///
+  /// If the [TIdentifier] is not present in the cache the value returned by the
+  /// given [valueFactory] is added to the cache and returned.
+  ///
+  /// All calls to [getAsync] await a call to the current caching strategy's
+  /// [CachingStrategy.onDidGet] method before returning the cached value. If the
+  /// [Cache] does not contain an instance for the given [TIdentifier], the given
+  /// [valueFactory] is called and a [CacheContext] event is emitted on the
+  /// [didUpdate] stream. A call to [get] that returns a cached value does not
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: `[get]` --> `[getAsync]`

File: lib/src/common/cache/cache.dart [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+    });
+
+    return _cache[id];
+  }
+
+  Iterable<TIdentifier> get keys => _cache.keys;
+
+  /// Marks a [TIdentifier] [TValue] pair as eligable for removal.
+  ///
+  /// The dicision of whether or not to actually remove the value will be up to
```
Comment (smaifullerton-wk) [inline lib/src/common/cache/cache.dart]: dicision -> decision

File: lib/src/common/cache/cache.dart [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+      _didUpdateController.add(new CacheContext(id, value));
+      return value;
+    });
+
+    return _cache[id];
+  }
+
+  Iterable<TIdentifier> get keys => _cache.keys;
+
+  /// Marks a [TIdentifier] [TValue] pair as eligable for removal.
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: `eligable` --> `eligible`

File: lib/src/common/cache/cache.dart (hunk 18 above)
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: `dicision` --> `decision`

File: lib/src/common/cache/cache.dart [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+    return new Future.value();
+  }
+
+  void _throwWhenDisposed(String op) {
+    if (isOrWillBeDisposed) {
+      throw new StateError('Cannot $op when Cache is disposed');
+    }
+  }
+
+  Future<Iterable<TValue>> get values => Future.wait(_cache.values);
```
Comment (todbachman-wf) [inline lib/src/common/cache/cache.dart]: Should we group the getters together? We usually do...

File: lib/src/common/cache/cache.dart (hunk 16 above)
Comment (todbachman-wf) [reply lib/src/common/cache/cache.dart]: Turns out the completer is what causes us to wait on the onDidGet. Helps if you read the code. 😉

**Comment by smaifullerton-wk** [comment]
### +10

- [x] Open sourcing this is fine
- [x] CI passes
- [x] Try this in the internal Workiva app that it came from to ensure that it still functions properly 
- [x] Try this in the internal Workiva app that consumes the aforementioned internal Workiva app to ensure that it still functions properly

**Comment by smaifullerton-wk** [comment]
### QA +1

- [x] Testing instruction
- [x] Dev +1's
- [x] Dev/QA +10 with detail of what was tested
- [ ] Security review (if required)
- [x] Unit tests created/updated
- [x] All unit tests pass
